backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/KeyspaceFactory.java

```diff
@@ -35,27 +35,39 @@
     private static final String SYSTEM_SCHEMA = "system_schema";
     private static final String KEYSPACES = "keyspaces";
     private static final String KEYSPACE_NAME = "keyspace_name";
+    private static final int CACHE_REPLICATION_FACTOR = 1;
 
     public static void createKeyspace(ClusterConfiguration clusterConfiguration, Cluster cluster) {
         if (clusterConfiguration.shouldCreateKeyspace()) {
             doCreateKeyspace(clusterConfiguration, cluster);
+            doCreateCacheKeyspace(clusterConfiguration, cluster);
         }
     }
 
     private static void doCreateKeyspace(ClusterConfiguration clusterConfiguration, Cluster cluster) {
+        createKeyspace(cluster, clusterConfiguration.getKeyspace(),
+            clusterConfiguration.getReplicationFactor(),
+            clusterConfiguration.isDurableWrites());
+    }
+
+    private static void createKeyspace(Cluster cluster, String keyspace, int replicationFactor, boolean durableWrites) {
```

Comment:
please create a configuration structure for each keyspace, it will simplify the existing code
also, the cache keyspace has nothing to do in backends-common, please move it to the cassandra-blobstore or whatever module that would fit

Reply:
> please create a configuration structure for each keyspace, it will simplify the existing code

:+1: 

> also, the cache keyspace has nothing to do in backends-common, please move it to the cassandra-blobstore or whatever module that would fit

I would tend to disagree here: we leverage the Cassandra driver to create a new capability (keyspace available for caching implementations). I find the concept generic enough to make it a top level concern of backend-cassandra.

It could be used for instance to provide a better distributed Tika cache (for instance)

Also, to be moving this "cache keyspace" in another place, we would need to handle keyspace registration & initialisation in a (guice) modular way, which is IMO clearly overkill.

Reply:
so you would argue creating a keyspace in cassandra (whatever the purpose of it) requires to create cached keyspace too?

I think that backend modules aim at providing tools, not taking decision about how to use these tools.

Reply:
> I think that backend modules aim at providing tools, not taking decision about how to use these tools.

Your descrition sounds like "the driver APIs are not nice enough so we provide our  layers around it".

Well, for me it aims at providing tools adapted to the usages we get in James.

Maybe I over-anticipate that "cache" need, but once other needs that CassandraDumbBlobStoreCache will arise, having the "cache session" being part of the backend would allow reuse without coupling.

Reply:
Though I agree that maybe we need to better wrap things with init operations on guice side too

Let me give it a try.

Reply:
> Your descrition sounds like "the driver APIs are not nice enough so we provide our layers around it".

I agree.

Just look at what is in this module:

* a way to describe our UDT and tables => not included in the driver
* some configuration helpers to expose only the parameters we want in James configuration
* a codec for ZonedDateTime => not included in the driver
* a tool to create keyspaces => not included in the driver
* a workaround to the lack of connection retry on start => not included in the driver
* a migration tool => not included in the driver
* a reactive wrapper for queries => not included in the driver
* healthcheck => not included in the driver

We have one thing wrong: in configuration, we have things related to specific cassandra tables and we shouldn't.

Everything else are just tools to have a higher level API to implement things in domain modules.

If you want a reusable cache thing, you'd rather create a new maven module for that IMO.

Reply:
I just contributed this split for Cassandra session.

The only "cache specific thing remaining in backend-cassandra is the "injection key" `@Named(InjectionNames.CACHE)` to be used to provide a cache session.

I consider that providing yet-another-maven module for a one line constant is overkill (think to the poor guy running `maven release`...)

However the important split (helper VS domain logic) in configuration POJOs as well as in the helper them selves is done, and the tests are corrected accordingly.

(that was quite a bunch of work!)